src/graphs/Breadth_First_Search.re

```diff
@@ -0,0 +1,91 @@
+type node = {
+    id: string,
+    neighbours: list(string),
+};
+type returnResult = {
+    level: Hashtbl.t(string, int), 
+    parent: Hashtbl.t(string, string)
+};
+
+exception NodeNotExist(string);
+
+let toHashTbl = adj_list => {
+    let adj_tbl = Hashtbl.create(List.length(adj_list));
+    let insert = node => {
+        Hashtbl.add(adj_tbl, node.id, node.neighbours);
+    };
+    let validateNeighbours = node => {
+        List.iter(neighbour => {
+            switch (Hashtbl.find(adj_tbl, neighbour)) {
+            | exception Not_found => raise(NodeNotExist(neighbour));
+            | _ => ();
+            };
+        }, node.neighbours);
+    };
+
+    List.iter(insert, adj_list);
+    List.iter(validateNeighbours, adj_list);
+
+    adj_tbl;
+};
+
+let visit = (node_id, ~adj_tbl, ~level, ~parent,~current_level) => {
+    switch (Hashtbl.find(level, node_id)) {
+    | exception Not_found => {
+        Hashtbl.add(level, node_id, current_level);
+        let neighbours = Hashtbl.find(adj_tbl, node_id);
+        let markParent = neighbour_id => {
+            switch (Hashtbl.find(parent, neighbour_id)) {
+                | exception Not_found => Hashtbl.add(parent, neighbour_id, node_id);
+                | _ => ();
+            }
+        };
+        List.iter(markParent, neighbours);
+        neighbours;   
+    }
+    | _ => ([]);
+    };
+};
+
+let rec traverse = (frontier, adj_tbl, level, parent, current_level) => {
```

Comment:
since visit uses named arguments why not traverse?

Reply:
I get issues when currying the visit function unless I explicitly name the arguments. I did not need to curry traverse, hence no named arguments. 
Should a name the traverse arguments anyway?

Reply:
traverse takes many arguments. It's better if they're named.